src/ansi-c/goto-conversion/goto_clean_expr.cpp

```diff
@@ -440,9 +476,155 @@ goto_convertt::clean_expr_resultt goto_convertt::clean_expr(
   }
   else if(expr.id() == ID_forall || expr.id() == ID_exists)
   {
+    quantifier_exprt &qex = to_quantifier_expr(expr);
+    code_expressiont code{qex.where()};
     DATA_INVARIANT(
-      !has_subexpr(expr, ID_side_effect),
-      "the front-end should check quantified expressions for side-effects");
+      !has_subexpr(expr, ID_side_effect) ||
+        (code.operands()[0].id() == ID_side_effect &&
+         code.operands()[0].get_named_sub()[ID_statement].id() ==
+           ID_statement_expression),
+      "quantifier must not contain side effects");
```

Comment:
Hmm, this one permits arbitrary side effects as long as they are encapsulated in a statement expression. IMHO the only permissible side effect should be statement expressions, so this check should apply recursively. Please use a lambda as suggested for the function-call test above.

Reply:
My idea in this PR is to convert all the side-effect expressions (except for function calls) in the quantifier body first, and then check that if the expression is side-effect free by checking if all the lhs of assigns are local declared. Let me double check if there could be other side-effect expression in MIR.